deeplearning4j-core/src/main/java/org/deeplearning4j/nn/layers/BaseLayer.java

```diff
@@ -534,18 +526,25 @@ public String toString() {
 
     @Override
     public Layer transpose() {
+        if(!(conf.getLayer() instanceof org.deeplearning4j.nn.conf.layers.FeedForwardLayer))
+            throw new UnsupportedOperationException("unsupported layer type: " + conf.getLayer().getClass().getName());
+
         INDArray W = getParam(DefaultParamInitializer.WEIGHT_KEY);
         INDArray b = getParam(DefaultParamInitializer.BIAS_KEY);
-        Layer layer = null;
+        Layer layer;
         try {
             Constructor c = getClass().getConstructor(NeuralNetConfiguration.class, INDArray.class, INDArray.class, INDArray.class);
-            NeuralNetConfiguration clone = conf.clone();
-            int nIn = clone.getNOut(),nOut = clone.getNIn();
-            clone.setNIn(nIn);
-            clone.setNOut(nOut);
+            NeuralNetConfiguration clone = conf.clone();  // assume a deep clone here
```

Comment:
This is the line that precipitated a move to deep cloning, since the nIn/nOut are now buried inside the layer instance.